### PR TITLE
chore: improve READMEs

### PR DESCRIPTION
**Pull request by vanbasten17**
Improve READMEs. https://linear.app/hubtype/issue/BOT2-17/add-examples-readme-and-explanation-in-their-website

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,79 @@
...
+| **[NLU](https://github.com/hubtype/botonic-examples/tree/master/templates/nlu)**                           | Starter template for your project with NLU capabilities.                                                                                                                                                                                                                                                                                                                                                                      |
+
+## Requirements
+
+- Node.js version 10 or higher
+- [NPM cli](https://docs.npmjs.com/cli/npm) or [Yarn](https://yarnpkg.com/en/)
+
+## Contributing with new examples
```
Comment (elozano98) [inline README.md:107]: Really good idea! :dart:

File: examples/nlu-assistant/README.md [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+    - [2.4 Using your own custom model](#24-using-your-own-custom-model)
+      - [2.4.1 Initializing the NLU with custom preprocessors](#241-initializing-the-nlu-with-custom-preprocessors)
+      - [2.4.2 Preparing the data](#242-preparing-the-data)
+      - [2.4.3 Model configuration and tunning](#243-model-configuration-and-tunning)
+      - [2.4.4 Training and evaluating the model](#244-training-and-evaluating-the-model)
+    - [2.5 Run the model](#25-run-the-model)
+  - [3. Defining the conversational flow](#3-defining-the-conversational-flow)
+  - [4. Using the Botonic NLU plugin](#4-using-the-botonic-nlu-plugin)
+
+## How to use this example
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: In the README view, hashtags are seen.

File: examples/nlu-assistant/README.md [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+
+1. From your command line, download the example by running:
+   ```bash
+   $ botonic new myNluAssistant nlu-assistant
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: What about replacing the myNluAssistant for `<BotName>`?

File: examples/nlu-assistant/README.md [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+get directions to the nearest ATM
+how do I get home taking E120
+find the fastest route to Baxter Building 42nd street Madison Avenue Manhattan
+how to get to the Main Street avoiding heavy traffic
+...
+```
+
+Each example should be defined in single line and should be separated of the next one by a newline.
+
+> **Note:** Take into account that in order to train deep learning models, the more example the better your neural network will perform. For the sake of this example we have added 50 examples per intent, but they could be insufficient.
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: I think it would be better to say:

>**Note:** Take into account that in Deep Learning models, the more examples you train with, the better the neural network will perform. For the sake of this example, we have added 50 examples per intent (but they could be insufficient).

File: examples/nlu-assistant/README.md [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+2. Now we will initialize an instance of `BotonicNLU` and feed it with our data:
+   ```ts
+   const nlu = new BotonicNLU()
+   const data = nlu.readData({
+     path: UTTERANCES_PATH,
+     language: LANGUAGE,
+     maxSeqLen: 20,
+   })
+   ```
+   `maxSeqLen` defines the maximum length of the sentences to train the NN. As webchat is a messaging channel, a `maxSeqLen` of 20 will be enough to cover our needs (_it's pretty uncommon that a user types an input of more of 20 words_).
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: maxSeqLen do not represent a limitation for the length of the sentences, it represents how long will be the sequence inputs:
>`maxSeqLen` defines the maximum length of the sequences generated to train the NN.

File: examples/nlu-assistant/README.md [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+3. Before start training our model, we need to split the data into two sets; one for training and one for evaluating the model. `testPercentage` is the percentage of examples that will be used to do the evaluation. This can be done with:
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: Maybe move the `testPercentage` explaination below the code snipet, as previously done for the `maxSeqLen`.

File: examples/nlu-assistant/README.md [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+   **Output:**
+
+   ```
+   ...
+   3283ms 15198us/step - acc=1.00 loss=8.80e-4 val_acc=0.167 val_loss=6.70
+   Accuracy: 0.6333333333333333
+   Model saved.
+   ```
+
+   As you can see, the `acc` (training accuracy) is high, and `val_acc` (validation accuracy) is low. This means that our network did a good job classifying the intents that it already knows, but it's performance when seeing new examples is really poor.
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: I would change it a little bit:
>As you can see, the `acc` (training accuracy) is high, and `val_acc` (validation accuracy) is low. This means that our network learned to correctly classify the inputs that it has already seen during the training, but the performance when seeing new examples is really poor. This problem is also known as **Overffiting**.

File: examples/nlu-assistant/README.md [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+   We could try to do lots of changes within this code, but we will never obtain significantly better results. To address this problem we will need to create a "dummier" model to be more flexible in terms of learning and correct this issue (known as `overfitting`), so we will need to make use of a custom NLU model.
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: I would not downgrade our SimpleNN template, instead, I would use this opportunity to present the other way of implementing a NN: **Creating your own model**:

>To address this problem we are going to try the other way of implementing a model. This option offers the possibility of creating a custom model without using any template.

File: examples/nlu-assistant/README.md [hunk 9]
```diff
@@ -0,0 +1,515 @@
...
+
+##### 2.2 Creating your own custom model
+
+In this example, we are going to create a common model's configuration in terms of natural language processing. This is a stack of `sequential` layers consisting of an `embedding` layer, followed by a `LSTM` layer and finally a `dense` layer.
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: I would try to explain it in an easier way:
>In this example, we are going to create a common model's configuration in terms of natural language processing. The sequential model is composed of three different layers: an **Embedding** layer, an **LSTM** (Long Short-Term Memory) layer, and finally a **Dense** layer with a `softmax` activation function to compute all intent probabilities.

File: examples/nlu-assistant/README.md [hunk 10]
```diff
@@ -0,0 +1,515 @@
...
+      layers.embedding({
+        inputDim: wordEmbeddingsMatrix.shape[0],
+        outputDim: wordEmbeddingsMatrix.shape[1],
+        inputLength: maxSeqLen,
+        trainable: false,
+        weights: [wordEmbeddingsMatrix],
+      })
+    )
+    ```
+    > **Note:** It's very important you define correctly these parameters for a correct functioning of the model.
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: In the README view, all of these parameter explanations are not in a list format.

File: examples/nlu-assistant/README.md [hunk 11]
```diff
@@ -0,0 +1,515 @@
...
+    > **`inputLength`**: must be the `maxSeqLen`.
+    > **`trainable`**: whether to freeze the weights of WE or leave them to variate. If you want to take all the profit of word embeddings, set it to `false`.
+    > **`weights`**: Loaded word embeddings matrix.
+3.  Add the rest of layers:
+    ```ts
+    model.add(layers.lstm(layersConfig.LSTM))
+    model.add(layers.dense(layersConfig.DENSE))
+    ```
+4.  Configure and prepare the model for training and evaluation.
+    `ts model.compile(modelCompileArgs) model.summary()`
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: I think you wanted to put this into a snippet code.

File: examples/nlu-assistant/README.md [hunk 12]
```diff
@@ -0,0 +1,515 @@
...
+
+##### 2.3 Creating your own custom data preprocessors
+
+In certain situations, you would want to tell `BotonicNLU` how to preprocess the data (removing accents, question marks, ...). To do so, `BotonicNLU` let's you to inject your own `tokenizer`, `normalizer` or `stemmer`. In this example we are going to use a custom `tokenizer` and `normalizer`. In order to maintain the project cleaner, we have created a `preprocessing-tools` directory in **src/nlu/**.
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: It is not exclusive, a user can inject the three of them:
>To do so, `BotonicNLU` let's you to inject your own `tokenizer`, `normalizer` and/or `stemmer`.

File: examples/nlu-assistant/README.md [hunk 13]
```diff
@@ -0,0 +1,515 @@
...
+
+**src/nlu/preprocessing-tools/tokenizer.ts**
+
+```ts
+const TreebankWordTokenizer = require('natural/lib/natural/tokenizers/treebank_word_tokenizer')
+
+export const tokenizer = new TreebankWordTokenizer()
+```
+
+In this example, we will be using [`natural`'s](https://github.com/NaturalNode/natural#tokenizers) `TreebankWordTokenizer` which works well for every language. To define your tokenizer, the only requirement is to export a class that implements a `tokenize` method.
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: Only natural linked:

> [`natural`](https://github.com/NaturalNode/natural#tokenizers)'s

File: examples/nlu-assistant/README.md [hunk 14]
```diff
@@ -0,0 +1,515 @@
...
+
+> **Note:** As `natural` is a library centered on Node and we need to have the bot working in a browser environment, we have to require the class explicitly from the absolute path.
+
+###### 2.3.2 Creating a custom normalizer
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: In the README view, hashtags are seen.

File: examples/nlu-assistant/README.md [hunk 15]
```diff
@@ -0,0 +1,515 @@
...
+}
+```
+
+In this normalizer, we are removing special characters we don't want to trait when the user enters an input. As well as `tokenizer`, this class must implement a `normalize` method.
+
+##### 2.4 Using your own custom model
+
+Now, let's move again to **src/nlu/train-model.ts**.
+
+###### 2.4.1 Initializing the NLU with custom preprocessors
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: In the README view, hashtags are seen.

File: examples/nlu-assistant/README.md [hunk 16]
```diff
@@ -0,0 +1,515 @@
...
+
+const nlu = new BotonicNLU({
+  tokenizer: tokenizer,
+  normalizer: new DefaultNormalizer(),
+})
+```
+
+Call `BotonicNLU` constructor and initialize it by passing your imported tokenizer and normalizer.
+
+###### 2.4.2 Preparing the data
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: In the README view, hashtags are seen.

File: examples/nlu-assistant/README.md [hunk 17]
```diff
@@ -0,0 +1,515 @@
...
+const [xTrain, xTest, yTrain, yTest] = nlu.trainTestSplit({
+  data: data,
+  testPercentage: 0.15,
+  stratify: false,
+})
+```
+
+At this point, you only had to call `readData` and all the preprocessing will be handled internally with the preprocesors you have injected before. As we have an homogenized distribution of the data (same number of examples per intent), we will be setting `stratify` option to `false`.
+
+###### 2.4.3 Model configuration and tunning
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: In the README view, hashtags are seen.

File: examples/nlu-assistant/README.md [hunk 18]
```diff
@@ -0,0 +1,515 @@
...
+```
+
+Time to configure our model.
+
+- **Word Embeddings**:
+  - Here we have chosen [`glove`](https://nlp.stanford.edu/projects/glove/) `50d` pretrained word embeddings. We must pass here the vocabulary extracted from previous data processing.
+- **LSTM Layer**:
+  - As we said before, we want our model to be less rigid when predicting over new inputs. To achive that, we use a technique called `regularization` that can be done by specifying `dropout` parameters to this layer. Also, reducing the number of `units` (neurons per layer) will help.
+- **Dense Layer**:
+  - This is the very last layer and **must** have as many units as intents we are targeting. Since we have multiple intents (classes) we have to make use of `softmax` activation.
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: The softmax activation function is not used because we have several intents. It is used to compute the probabilities of all intents so the sum of all of them is equal to 1.

File: examples/nlu-assistant/README.md [hunk 19]
```diff
@@ -0,0 +1,515 @@
...
+
+###### 2.4.4 Training and evaluating the model
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: In the README view, hashtags are seen.

File: examples/nlu-assistant/README.md [hunk 20]
```diff
@@ -0,0 +1,515 @@
...
+
+Now we have to add the compilation parameters for our model, this is: choosing an `optimizer`, a `loss` (`sparseCategoricalCrossentropy` for multiple classes) and the `metrics` we want to track. Here the most obvious option is to use the `accuracy` metric.
+Finally, we put all the pieces together so we:
+
+1. Assign our model to current `nlu` instance with `nlu.model`.
+2. Do iterative cycles of training during the defined number of `epochs`.
+3. Save the model.
+4. Evaluate it.
+
+##### 2.5 Run the model
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: In the README view, hashtags are seen.

File: examples/nlu-assistant/README.md [hunk 21]
```diff
@@ -0,0 +1,515 @@
...
+1979ms 9703us/step - acc=0.897 loss=0.261 val_acc=0.961 val_loss=0.154
+Accuracy: 0.8888888888888888
+Model saved.
+```
+
+As you can see, we have been able to increase our accuracy up to a 88% and have more closer values between `acc` and `val_acc` which are good signs that our model is working properly. Also the values for both losses are decreasing over every training epoch. So we are now ready to define our conversational flow!
+
+> **Note:** These results are not deterministic because NN weights are initialized with random values, so maybe you have to run the `botonic train` script various times before obtaining a convincing result.
+
+#### 3. Defining the conversational flow
```
Comment (elozano98) [inline examples/nlu-assistant/README.md]: In the README view, hashtags are seen.

**Review by elozano98 — CHANGES_REQUESTED** [review]
Some technical explanations of the NLU README are needed to be modified but you did a **really good job**! :dart:

**Comment by vanbasten17** [comment]
Amazing review @elozano98 !! thanks a lot!! 🚀
